File: app/src/main/java/com/paragon/sensonic/ui/activities/login/LoginViewModel.java
package com.paragon.sensonic.ui.activities.login;

import android.content.Context;
import android.text.TextUtils;
import android.util.Patterns;

import com.paragon.sensonic.R;
import com.paragon.sensonic.auth.PasswordLessLoginResponse;
import com.paragon.sensonic.auth.dto.PasswordLessLogin;
import com.paragon.utils.GeneralFunctions;
import com.paragon.utils.base.BaseViewModel;
import com.paragon.utils.networking.NetworkResponseCallback;

import java.util.HashMap;

public class LoginViewModel extends BaseViewModel<LoginNavigator> {

    public void init() {
        getNavigator().init();
    }

    public void onLoginButtonClick() {
        getNavigator().onLoginClick();
    }

    public void onCountyCodeClick(){
        getNavigator().onCountyCodeClick();
    }

    public void onMobileClick() {
        getNavigator().onMobileClick();
    }

    public void onEmailClick() {
        getNavigator().onEmailClick();
    }

    private boolean isValid(Context context, String code, String value, String isMobile) {
        if (isMobile.matches("phone")) {
            if (TextUtils.isEmpty(value)) {
                getNavigator().setErrorText(true, context.getString(R.string.error_empty_mobile_number));
                return false;
            } else if (!GeneralFunctions.isValidPhoneNumber(value)) {
                getNavigator().setErrorText(true, context.getString(R.string.error_invalid_mobile_number));
                return false;
            } else if(!GeneralFunctions.validateUsing_libphonenumber(code,value)){
                getNavigator().setErrorText(true, context.getString(R.string.error_invalid_mobile_number));
                return false;
            }
            else {
                getNavigator().setErrorText(false, "");
                return true;
            }
        } else {
            if (TextUtils.isEmpty(value.toString())) {
                return false;
            } else if (!Patterns.EMAIL_ADDRESS.matcher(value).matches()) {
                return false;
            } else {
                return true;
            }
        }
    }

    public void callLoginApi(Context context, String code, String value, String type) {
        if (isValid(context, code, value, type)) {
            getNavigator().onShowProgress();
            HashMap<String, String> data = new HashMap<>();
            data.put("type",type);
            if(TextUtils.isEmpty(code)) {
                data.put(type, value);
            }else{
                data.put(type, code+"-"+value);
            }

            new PasswordLessLoginResponse().doNetworkRequest(data, new NetworkResponseCallback<PasswordLessLogin>() {
                @Override
                public void onResponse(PasswordLessLogin object) {
                    getNavigator().onHideProgress();
                    getNavigator().onSuccess(object);
                }

                @Override
                public void onFailure(String error) {
                    getNavigator().onHideProgress();
                    getNavigator().onError(error);
                }

                @Override
                public void onInternetDisable() {

                }
            });
        }
    }
}

File: app/src/main/java/com/paragon/sensonic/ui/activities/login/LoginViewModel.kt
package com.paragon.sensonic.ui.activities.login

import android.text.TextUtils
import android.util.Patterns
import com.paragon.sensonic.App
import com.paragon.sensonic.R
import com.paragon.sensonic.auth.PasswordLessLoginResponse
import com.paragon.sensonic.auth.RefreshTokenResponse
import com.paragon.sensonic.auth.dto.Credentials
import com.paragon.sensonic.auth.dto.PasswordLessLogin
import com.paragon.utils.GeneralFunctions
import com.paragon.utils.base.BaseViewModel
import com.paragon.sensonic.utils.local.AppPreference
import com.paragon.sensonic.utils.local.PreferenceKeys
import com.paragon.utils.networking.NetworkResponseCallback
import kotlin.collections.HashMap

class LoginViewModel : BaseViewModel<LoginNavigator>() {

    fun init() {
        navigator.init()
    }

    fun onLoginButtonClick() {
        navigator.onLoginClick()
    }

    fun onCountyCodeClick() {
        navigator.onCountyCodeClick()
    }

    fun onMobileClick() {
        navigator.onMobileClick()
    }

    fun onEmailClick() {
        navigator.onEmailClick()
    }

    private fun isValid(code: String, value: String, isMobile: String): Boolean {
        return if (isMobile.equals("phone")) {
            if (TextUtils.isEmpty(value)) {
                navigator.setErrorText(
                    true,
                    App.getAppContext().getString(R.string.error_empty_mobile)
                )
                false
            } else if (!GeneralFunctions.isValidPhoneNumber(value)) {
                navigator.setErrorText(
                    true,
                    App.getAppContext().getString(R.string.error_incorrect_mobile)
                )
                false
            } else if (!GeneralFunctions.validateUsing_libphonenumber(code, value)) {
                navigator.setErrorText(
                    true,
                    App.getAppContext().getString(R.string.error_incorrect_mobile)
                )
                false
            } else {
                navigator.setErrorText(false, "")
                true
            }
        } else {
            if (TextUtils.isEmpty(value)) {
                navigator.setErrorText(
                    true,
                    App.getAppContext().getString(R.string.error_empty_email)
                )
                false
            } else if (!Patterns.EMAIL_ADDRESS.matcher(value).matches()) {
                navigator.setErrorText(
                    true,
                    App.getAppContext().getString(R.string.error_incorrect)
                )
                false
            } else {
                navigator.setErrorText(false, "")
                true
            }
        }
    }

    fun callLoginApi(code: String, value: String, type: String) {
        if (isValid(code, value, type)) {
            navigator.onShowProgress()
            val data = HashMap<String, String>()
            data["type"] = type
            if (TextUtils.isEmpty(code)) {
                data[type] = value
            } else {
                data[type] = "$code-$value"
            }
            PasswordLessLoginResponse().doNetworkRequest(
                data, object : NetworkResponseCallback<PasswordLessLogin> {

                    override fun onResponse(response: PasswordLessLogin) {
                        navigator.apply {
                            onHideProgress()
                            onSuccess(response)
                        }
                    }

                    override fun onFailure(error: String) {
                        navigator.apply {
                            onHideProgress()
                            if (error == App.getAppContext().getString(R.string.invalid_user_name)) {
                                if (type == "email") {
                                    onError(App.getAppContext().getString(R.string.error_invalid_email))
                                } else {
                                    onError(App.getAppContext().getString(R.string.error_not_register_mobile))
                                }
                            }
                        }
                    }

                    override fun onInternetDisable() {}

                })
        }
    }


}